### PR TITLE
Extend barbery for using non linear structure in storage and cache

### PR DESCRIPTION
**Pull request by lionjsa**
also added 2 new commits:
- checking extension openssl are loaded
- suggestion for using openssl extension in composer.json

### CODE REVIEW

File: library/Barberry/Uniq.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+class Uniq
+{
+    public static function id($len = 10)
+    {
+        if (extension_loaded('openssl')) {
+            $bytes = openssl_random_pseudo_bytes($len);
+            return bin2hex($bytes);
+        }
+        return $len > 10 ? md5(uniqid('', true)) : uniqid('');
```
Comment (Magomogo) [inline library/Barberry/Uniq.php]: please make it a function.

File: library/Barberry/Storage/File.php [hunk 2]
```diff
@@ -51,21 +60,26 @@ public function delete($id) {
...
         }
     }
 
-//--------------------------------------------------------------------------------------------------
-
     private function generateUniqueId() {
-        $tempFile = tempnam($this->permanentStoragePath, '');
-        chmod($tempFile, 0664);
-        return basename($tempFile);
+        $destination = NonLinearDestination::factory($uId = Uniq::id())->make($this->permanentStoragePath, self::DIR_MODE);
```
Comment (Magomogo) [inline library/Barberry/Storage/File.php]: it is make sense to delegate collision check to NonLinearDestination class

File: library/Barberry/Storage/File/NonLinearDestination.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        return $this->baseLen > 10 ? md5(uniqid('', true)) : uniqid('');
+    }
+
+    /**
+     * Append last slash
+     *
+     * @param string $path
+     * @return string
+     */
+    public static function als($path)
```
Comment (Magomogo) [inline library/Barberry/Storage/File/NonLinearDestination.php]: Why this function is part of public interface of `NonLinearDestination` object?

File: library/Barberry/Storage/File/NonLinearDestination.php [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        $hash = $this->generateHash();
+        $start = 0;
+        $d = $this->depth;
+        $dir = array();
+
+        while ($d-- > 0) {
+            $dir[] = substr($hash, $start, $this->len);
+            $start += $this->len;
+        }
+        return $this->basePath . self::als(implode(DIRECTORY_SEPARATOR, $dir));
```
Comment (Magomogo) [inline library/Barberry/Storage/File/NonLinearDestination.php]: What about collision protection?

File: library/Barberry/Storage/File/NonLinearDestination.php [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+    private $depth = 3;
+
+    private $len = 2;
+
+    private $basePath;
+
+    private $baseLen = 10;
+
+    private function __construct($basePath, $base)
```
Comment (Magomogo) [inline library/Barberry/Storage/File/NonLinearDestination.php]: What is `$base` for?

File: library/Barberry/Storage/File/NonLinearDestination.php (hunk 4 above)
Comment (lionjsa) [inline library/Barberry/Storage/File/NonLinearDestination.php]: see getBase method

File: library/Barberry/Storage/File/NonLinearDestination.php (hunk 3 above)
Comment (lionjsa) [inline library/Barberry/Storage/File/NonLinearDestination.php]: also used in Cache class

File: library/Barberry/Storage/File/NonLinearDestination.php (hunk 5 above)
Comment (lionjsa) [inline library/Barberry/Storage/File/NonLinearDestination.php]: see factory() method phpdoc

File: library/Barberry/Storage/File.php [hunk 6]
```diff
@@ -1,12 +1,18 @@
 <?php
 namespace Barberry\Storage;
 
+use function Barberry\file\als;
```
Comment (Magomogo) [inline library/Barberry/Storage/File.php]: not supported by PHP 5.3, please replace.

File: test/integration/CacheTest.php [hunk 7]
```diff
@@ -16,42 +18,30 @@ protected function tearDown() {
...
     public function testIsContentSavedInFileSystem() {
-        $this->cache()->save(
+        $id = $this->cache()->save(
             Test\Data::gif1x1(),
             new Request('/7yU98sd_1x1.gif')
         );
 
-        $expectedPath = $this->cache_path . '/7yU98sd/7yU98sd_1x1.gif';
+        $path = $this->cache_path . generate($id);
+        $expectedPath = $path . '/7yU98sd/7yU98sd_1x1.gif';
```
Comment (Magomogo) [inline test/integration/CacheTest.php]: temporary variable `$path` makes this difficult to read

File: library/Barberry/Cache.php [hunk 8]
```diff
@@ -1,50 +1,51 @@
...
+use function Barberry\file\als;
+
 class Cache {
 
     private $path;
 
     public function __construct($path) {
-        $this->path = rtrim($path,'/') . '/';
+        $this->path = als($path);
+        set_error_handler(array($this, 'errorHandler'));
```
Comment (Magomogo) [inline library/Barberry/Cache.php]: It's not a good idea to use error handlers in libraries, we should honor global error handlers set by the top level application.

File: library/Barberry/Storage/File.php [hunk 9]
```diff
@@ -51,21 +55,37 @@ public function delete($id) {
...
+    /**
+     * @param $id
+     * @return string
+     */
+    private function filePathById($id) {
+        if (is_file($f = $this->permanentStoragePath . $id)) {
+            return $f;
+        }
+
+        return $this->permanentStoragePath . generate($id) . $id;
```
Comment (Magomogo) [inline library/Barberry/Storage/File.php]: Bad naming: It's hard to understand `generate` out of context.

File: library/Barberry/Storage/File.php [hunk 10]
```diff
@@ -1,12 +1,18 @@
...
 class File implements StorageInterface {
 
     private $permanentStoragePath;
 
+    private $baseLen = 10;
+
     public function __construct($path) {
-        $this->permanentStoragePath = rtrim($path, '/') . '/';
+        $this->permanentStoragePath = als($path);
+        set_error_handler(array($this, 'errorHandler'));
```
Comment (Magomogo) [inline library/Barberry/Storage/File.php]: Error handlers are global. When you set one you disable previous.

File: test/integration/CacheTest.php [hunk 11]
```diff
@@ -16,42 +18,30 @@ protected function tearDown() {
...
         $this->assertEquals(file_get_contents($expectedPath), Test\Data::gif1x1());
     }
 
-//--------------------------------------------------------------------------------------------------
-
     private function cache() {
         return new Cache($this->cache_path);
     }
-
-    private static function rmDirRecursive($dir) {
```
Comment (Magomogo) [inline test/integration/CacheTest.php]: please move this function under `fs` namespace.

**Pull request merged**